File: app/src/main/java/com/example/myapplication/data/WorkoutRepository.kt
package com.example.myapplication.data

import com.example.myapplication.core.model.ActiveGoal
import com.example.myapplication.core.model.CompletedWorkout
import com.example.myapplication.core.model.GoalConfig
import com.example.myapplication.core.model.ProgramTemplate
import com.example.myapplication.core.model.WorkoutSession
import kotlinx.coroutines.flow.Flow

interface WorkoutRepository {
    fun observeActiveGoal(): Flow<ActiveGoal?>
    fun observeCurrentWorkout(): Flow<WorkoutSession?>
    fun observeCompletedWorkouts(): Flow<List<CompletedWorkout>>
    suspend fun createGoal(config: GoalConfig, program: ProgramTemplate, startEpochDay: Long)
    suspend fun setExerciseChecked(sessionId: Long, orderIndex: Int, checked: Boolean)
    suspend fun completeWorkout(sessionId: Long, completedEpochDay: Long): CompleteWorkoutResult
    suspend fun substituteExercise(
        sessionId: Long,
        orderIndex: Int,
        replacementExerciseId: String,
    ): ExerciseSubstitutionResult = ExerciseSubstitutionResult.InvalidCandidate
    suspend fun applyTimeBudget(sessionId: Long, minutes: Int?): TimeBudgetResult =
        TimeBudgetResult.StaleSession
    suspend fun archiveActiveGoal()
}

sealed interface TimeBudgetResult {
    data object Applied : TimeBudgetResult
    data object InvalidBudget : TimeBudgetResult
    data object HasCheckedExercises : TimeBudgetResult
    data object StaleSession : TimeBudgetResult
}

sealed interface ExerciseSubstitutionResult {
    data object Applied : ExerciseSubstitutionResult
    data object InvalidCandidate : ExerciseSubstitutionResult
    data object StaleSession : ExerciseSubstitutionResult
    data object AlreadyChecked : ExerciseSubstitutionResult
}

sealed interface CompleteWorkoutResult {
    data object Completed : CompleteWorkoutResult
    data object BlockedByUncheckedExercises : CompleteWorkoutResult
    data object AlreadyCompleted : CompleteWorkoutResult
}

File: flutter/lib/data/repositories/workout_repository.dart
import '../../../core/model/achievement_models.dart';
import '../../../core/model/goal_models.dart';
import '../../../core/model/workout_models.dart';
import '../../../core/model/catalog_models.dart';
import '../../../core/program/schedule_rescheduler.dart';

enum CompleteWorkoutResult {
  completed,
  blockedByUncheckedExercises,
  alreadyCompleted,
}

enum ExerciseSubstitutionResult {
  applied,
  invalidCandidate,
  staleSession,
  alreadyChecked,
}

enum TimeBudgetResult {
  applied,
  invalidBudget,
  hasCheckedExercises,
  staleSession,
}

enum ScheduleChangeResult {
  applied,
  stale,
}

abstract class WorkoutRepository {
  Stream<ActiveGoal?> observeActiveGoal();
  Stream<WorkoutSession?> observeCurrentWorkout();
  Stream<List<CompletedWorkout>> observeCompletedWorkouts();
  Stream<List<WorkoutHistoryEntry>> observeWorkoutHistory();
  Future<void> createGoal(
      GoalConfig config, ProgramTemplate program, int startEpochDay);
  Future<void> setExerciseChecked(int sessionId, int orderIndex, bool checked);
  Future<CompleteWorkoutResult> completeWorkout(
      int sessionId, int completedEpochDay);
  Future<ExerciseSubstitutionResult> substituteExercise({
    required int sessionId,
    required int orderIndex,
    required String replacementExerciseId,
  });
  Future<TimeBudgetResult> applyTimeBudget(int sessionId, int? minutes);
  Future<ScheduleChangePreview> previewScheduleChange(
      int sessionId, int newEpochDay);
  Future<ScheduleChangeResult> applyScheduleChange(
      ScheduleChangePreview preview);
  Future<void> archiveActiveGoal();

  /// Các badge đã mở khoá và được lưu lại từ trước.
  Future<Set<AchievementType>> unlockedAchievements();

  /// Lưu các badge vừa mở khoá. Ghi trùng là vô hại (insertOrIgnore).
  Future<void> recordUnlockedAchievements(
      Set<AchievementType> types, int unlockedAtEpochMillis);
}
